asapdiscovery-data/asapdiscovery/data/schema_v2/schema_base.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import annotations
+
+import json
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, Optional, Tuple, Union  # noqa: F401
+
+from pydantic import BaseModel, ByteSize, validator
+
+
+class DataStorageType(str, Enum):
+    sdf = "sdf"
+    pdb = "pdb"
+
+
+def read_file_directly(file: str | Path) -> str:
+    with open(str(file)) as f:
+        contents = f.read()
+    return contents
+
+
+def write_file_directly(file: str | Path, data: str) -> None:
+    with open(str(file), "w") as f:
+        f.write(data)
+
+
+def utf8len(s: str) -> int:
+    return len(s.encode("utf-8"))
+
+
+class DataModelAbstractBase(BaseModel):
+    """
+    Base class for asapdiscovery pydantic models that simplify dictionary, JSON
+    and other behaviour
+    """
+
+    @classmethod
+    def from_dict(cls, dict):
+        return cls.parse_obj(dict)
+
+    @classmethod
+    def from_json(cls, json_str):
+        return cls.parse_obj(json.loads(json_str))
+
+    # check_fields required to be false as `data` not defined for baseclass
+    @validator("data", check_fields=False)
+    def data_must_not_be_empty(cls, v):
```

Comment:
Pydantic has this validation built in, if you change any required fields to a default value of `...` (ellipsis) pydantic will raise a 'ValidationError' if you create a model with a missing required field.

Reply:
is there a way to make a method required in the base class but changeable in the subclasses? I think the problem here is that `data` is defined in the subclasses but not the base class. if we make it an abstract method and make it required in the base class will that requirement hold if we overwrite the method in subclasses?

Reply:
Yep no way to make it required but you just have base class raise NotImplementedError and force the child classes to implement it. 

I'll make the changes and push tommorow morn.

Reply:
I have just removed this as @jthorton suggests.